pull-requests/code-reviews/recipes/Python.md

```diff
@@ -1,53 +1,207 @@
+Install `Pylint`
 
-### [```Flake8```](https://pypi.org/project/flake8/)
+```bash
+pip install pylint
+```
+
+Run `Pylint`
+
+```bash
+pylint src  # lint the source directory
+```
 
-Flake8 is an option for teams who want a single wrapper for running Pyflakes and pycodestyle.
+## Automatically format python code
 
-## Unit Testing
+### Black
+
+[Black](https://github.com/psf/black) is an unappologetic code formatting tool. It removes all need from `pycodestyle` nagging about formatting so the team can focus on content vs style. It is however not possible to configure black for your own style needs.
+
+```bash
+pip install black
+```
+
+Format python code
+
+```bash
+black [file/folder]
+```
+
+### Autopep8
+
+[Autopep8](https://github.com/hhatto/autopep8) is more lenient and allows more configuration if you want less stringent formatting.
+
+```bash
+pip install autopep8
+```
 
-### [```Unittest``](https://docs.python.org/3/library/unittest.html)
-Unittest is a simple unit testing framework that provides the basics for unit testing: asserts, fixtures, mocks, etc. There is no need to install `unittest` and tests can be run using:
+Format python code
 
 ```bash
-python -m unittest
+autopep8 [file/folder] --in-place
 ```
 
-### [```Pytest```](https://docs.pytest.org/en/latest/)
-Pytest is a simple unit testing framework that provides the basics for unit testing: asserts, fixtures, etc... 
+### yapf
+
+[yapf](https://github.com/google/yapf) Yet Another Python Formatter is a python formatter from Google based on ideas from gofmt.  This is also more configurable and a good option for automatic code formatting.
 
-```bash 
-pip3 install pytest
+```bash
+pip install yapf
 ```
 
-### [```Pytest-mock```](https://github.com/pytest-dev/pytest-mock/)
-Mocking/stubbing framework that works with Pytest. Use this to mock out OS calls, off-box calls, and other calls that would unnecessarily broaden the scope of unit tests.
+Format python code
 
 ```bash
-pip3 install pytest-mock
+yapf [file/folder] --in-place
 ```
 
+## VS Code Extensions
+
+### Python
+
+The [Python language extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) is the base extension you should have installed for python development with VS Code. It enables among other things, intellisense, debugging, linting (with the above linters), testing with pytest or unittest, and code formatting with the formatters mentioned above.
+
+### Pyright
+
+The [Pyright extension](https://marketplace.visualstudio.com/items?itemName=ms-pyright.pyright) augments VS Code with static type checking when you use type hints
+
+```python
+def add(first_value: int, second_value: int) -> int:
+    return first_value + second_value
+```
+
+## Build validation
+
+To automate linting with `flake8` and testing with `pytest` in Azure Devops you can add the following snippet to you `azure-pipelines.yaml` file.
+
+```yaml
+trigger:
```

Comment:
I'm back and forth on how much we should include for YAML pipelines. Should we just focus on the actual linter tasks or does it make sense to include triggers here. More content can be easier to setup for ppl, but it's also more static content we need to maintain.

Reply:
I'll leave as is for now and we can have a discussion around this in the team checkpoint